# Analytics of submissions to popular technology site Hacker News

In this project, we're specifically interested in posts whose titles begin with either Ask HN or Show HN. 

* Users submit Ask HN posts to ask the Hacker News community a specific question.
* Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

We'll compare these two types of posts to determine the following:

1. Do Ask HN or Show HN receive more comments on average?
2. Do posts created at a certain time receive more comments on average?


___

#### We begin by importing the libraries needed and reading the data set into a list of lists.

In [1]:
from csv import reader

opened_file = open("hacker_news.csv")
read_file = reader(opened_file)
hn = list(read_file)
headers = hn[0]
hn = hn[1:]

print(headers)
print('\n')
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


Now that we've removed **the headers from hn, we're ready to filter our data**. 

Since we're only concerned with post titles beginning with Ask HN or Show HN, we'll create new lists of lists containing just the data for those titles.

To do this, 
1. we use the "str".startswith('str')
2. to check for cases we use "str".lower() method


In [2]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
            
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


##### Filtering data into **Three** main groups

1. Ask HN posts = 1744
2. Show HN posts = 1162
3. Other posts = 17194

Below are the first five rows in the ask_posts list of lists:

In [3]:
print(ask_posts[:5])

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


Below are the first five rows in the show_posts list of lists:

In [4]:
print(show_posts[:5])

[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 'h8liu', '4/28/2016 18:05'], ['12178806', 'Show HN: Webscope  Easy way for web developers to communicate with Clients', 'http://webscopeapp.com', '3', '3', 'fastbrick', '7/28/2016 7:11'], ['10872799', 'Show HN: GeoScreenshot  Easily test Geo-IP based web pages', 'https://www.geoscreenshot.com/', '1', '9', 'kpsychwave', '1/9/2016 20:45']]


Below are the first five rows in the other_posts list of lists:

In [5]:
print(other_posts[:5])

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


#### Next, let's determine if ask posts or show posts receive more comments on average.

In [6]:
# find the average number of commments 
# in the ask posts

total_ask_comments = 0

for row in ask_posts:
    num_comments = row[4]
    num_comments = int(num_comments)
    total_ask_comments += num_comments

# computing average number of comments
avg_ask_comments = total_ask_comments / len(ask_posts)

string = 'The average number of comments from the ask posts: {}'
string.format (avg_ask_comments)

'The average number of comments from the ask posts: 14.038417431192661'

In [7]:
# Finding the average number of comments 
# from the show posts

total_show_comments = 0

for row in show_posts:
    num_comments = row[4]
    num_comments = int(num_comments)
    total_show_comments += num_comments
    
# computing averagenumber of comments
avg_show_comments = total_show_comments / len(show_posts)

string_1 = 'The average number of comments from the show posts: {}'
string_1.format (avg_show_comments)

'The average number of comments from the show posts: 10.31669535283993'

* The Average number of comments received from ask posts is **14.038417431192661 whiles that of the show posts is 10.31669535283993.**

* Thus it is determined from the data that Ask posts receive more comments than show posts. 

* This can be attributed to the fact that many users tend to post answers to questions asked on the platform as compared to commenting on other posts

Next, we'll determine if ask posts created at a certain time are more likely to attract comments. We'll use the following steps to perform this analysis:

1. Calculate the amount of ask posts created in each hour of the day, along with the number of comments received.

2. Calculate the average number of comments ask posts receive by hour created.

In [8]:
# importing the datetime module as dt
# date format = %m/%d/%Y %H:%M

import datetime as dt

result_list = []

for row in ask_posts:
    created_at = row[6]
    num_comments = row[4]
    num_comments = int(num_comments)
    result_list.append([created_at, num_comments])
    
print(result_list[:5])
    

[['8/16/2016 9:55', 6], ['11/22/2015 13:43', 29], ['5/2/2016 10:14', 1], ['8/2/2016 14:20', 3], ['10/15/2015 16:38', 17]]


In [9]:
# calculating the number of posts per hour created 
# and comments per hour created

counts_by_hour = {}
comments_by_hour = {}

for pair in result_list:
    date = pair[0]
    date = dt.datetime.strptime(date, '%m/%d/%Y %H:%M')
    hour = dt.datetime.strftime(date, "%H")
    comments = pair[1]
    
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = comments
        
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += comments
    
        


In [10]:
print(counts_by_hour)

{'02': 58, '03': 54, '11': 58, '04': 47, '17': 100, '08': 48, '10': 59, '01': 60, '20': 80, '18': 109, '16': 108, '13': 85, '12': 73, '23': 68, '07': 34, '15': 116, '00': 55, '05': 46, '21': 109, '19': 110, '14': 107, '22': 71, '09': 45, '06': 44}


In [11]:
print(comments_by_hour)

{'02': 1381, '03': 421, '11': 641, '04': 337, '17': 1146, '08': 492, '10': 793, '01': 683, '20': 1722, '18': 1439, '16': 1814, '13': 1253, '12': 687, '23': 543, '07': 267, '15': 4477, '00': 447, '05': 464, '21': 1745, '19': 1188, '14': 1416, '22': 479, '09': 251, '06': 397}


Next, we'll use these two dictionaries to calculate the average number of comments for posts created during each hour of the day.

In [12]:
avg_by_hour = []

for hour in comments_by_hour:
    avg_by_hour.append([hour, comments_by_hour[hour] / counts_by_hour[hour]])

avg_by_hour

[['02', 23.810344827586206],
 ['03', 7.796296296296297],
 ['11', 11.051724137931034],
 ['04', 7.170212765957447],
 ['17', 11.46],
 ['08', 10.25],
 ['10', 13.440677966101696],
 ['01', 11.383333333333333],
 ['20', 21.525],
 ['18', 13.20183486238532],
 ['16', 16.796296296296298],
 ['13', 14.741176470588234],
 ['12', 9.41095890410959],
 ['23', 7.985294117647059],
 ['07', 7.852941176470588],
 ['15', 38.5948275862069],
 ['00', 8.127272727272727],
 ['05', 10.08695652173913],
 ['21', 16.009174311926607],
 ['19', 10.8],
 ['14', 13.233644859813085],
 ['22', 6.746478873239437],
 ['09', 5.5777777777777775],
 ['06', 9.022727272727273]]

Although we now have the results we need, this format makes it hard to identify the hours with the highest values. 

Let's finish by sorting the list of lists and printing the five highest values in a format that's easier to read.

In [13]:
# Create a list that equals avg_by_hour with swapped columns.

swap_avg_by_hour = []

for rows in avg_by_hour:
    hour = rows[0]
    avg = rows[1]
    swap_avg_by_hour.append([avg, hour])

swap_avg_by_hour

[[23.810344827586206, '02'],
 [7.796296296296297, '03'],
 [11.051724137931034, '11'],
 [7.170212765957447, '04'],
 [11.46, '17'],
 [10.25, '08'],
 [13.440677966101696, '10'],
 [11.383333333333333, '01'],
 [21.525, '20'],
 [13.20183486238532, '18'],
 [16.796296296296298, '16'],
 [14.741176470588234, '13'],
 [9.41095890410959, '12'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [38.5948275862069, '15'],
 [8.127272727272727, '00'],
 [10.08695652173913, '05'],
 [16.009174311926607, '21'],
 [10.8, '19'],
 [13.233644859813085, '14'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09'],
 [9.022727272727273, '06']]

The `sorted()` function is used to sort **swap_avg_by_hour** in descending order. 

Since the first column of this list is the average number of comments, sorting the list will sort by the average number of comments.

In [14]:
sorted_swap = sorted(swap_avg_by_hour, reverse = True)

sorted_swap

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

#### Top 5 Hours for Ask Posts Comments

In [15]:
for row in sorted_swap:
    hours = row[-1]
    hours = dt.datetime.strptime(hours, "%H")
    hours = dt.datetime.strftime(hours, "%H:%M")
    avg_posts = row[0]
    
    string_out = "{time}: {avg_num: .2f} average comments per post"
    output = string_out.format(time=hours,avg_num = avg_posts )
    print(output)

15:00:  38.59 average comments per post
02:00:  23.81 average comments per post
20:00:  21.52 average comments per post
16:00:  16.80 average comments per post
21:00:  16.01 average comments per post
13:00:  14.74 average comments per post
10:00:  13.44 average comments per post
14:00:  13.23 average comments per post
18:00:  13.20 average comments per post
17:00:  11.46 average comments per post
01:00:  11.38 average comments per post
11:00:  11.05 average comments per post
19:00:  10.80 average comments per post
08:00:  10.25 average comments per post
05:00:  10.09 average comments per post
12:00:  9.41 average comments per post
06:00:  9.02 average comments per post
00:00:  8.13 average comments per post
23:00:  7.99 average comments per post
07:00:  7.85 average comments per post
03:00:  7.80 average comments per post
04:00:  7.17 average comments per post
22:00:  6.75 average comments per post
09:00:  5.58 average comments per post


** Top 5 Hours for Ask Posts Comments **
1. 15:00 38.59 average comments per post
2. 02:00 23.81 average comments per post
3. 20:00 21.52 average comments per post
4. 16:00 16.80 average comments per post
5. 21:00 16.01 average comments per post


Thus the answer to the question:

>Which hours should you create a post during to have a higher chance of receiving comments?

The best time to create a post that would be most likely to receive more comments is at ** 15:00 (the time zone is Eastern Time in the US).**  An average of 38.59 comments per post is expected.

There's about a **60% increase** in the number of comments between the hours with the highest and second highest average number of comments.

Inorder to better use this information, we need to convert the time to GMT (Greenwich Mean Time). **Eastern Daylight Time is 4 hours behind of Greenwich Mean Time**


In [16]:
gmt_time_sorted = []
add_time = dt.timedelta(hours=4)

for row in sorted_swap:
    time = row[-1]
    time = dt.datetime.strptime(time, "%H")
    gmt_time = time + add_time
    gmt_time = dt.datetime.strftime(gmt_time, "%H:%M")
    avg_comm = row[0]
    gmt_time_sorted.append([gmt_time, avg_comm])
    
gmt_time_sorted

[['19:00', 38.5948275862069],
 ['06:00', 23.810344827586206],
 ['00:00', 21.525],
 ['20:00', 16.796296296296298],
 ['01:00', 16.009174311926607],
 ['17:00', 14.741176470588234],
 ['14:00', 13.440677966101696],
 ['18:00', 13.233644859813085],
 ['22:00', 13.20183486238532],
 ['21:00', 11.46],
 ['05:00', 11.383333333333333],
 ['15:00', 11.051724137931034],
 ['23:00', 10.8],
 ['12:00', 10.25],
 ['09:00', 10.08695652173913],
 ['16:00', 9.41095890410959],
 ['10:00', 9.022727272727273],
 ['04:00', 8.127272727272727],
 ['03:00', 7.985294117647059],
 ['11:00', 7.852941176470588],
 ['07:00', 7.796296296296297],
 ['08:00', 7.170212765957447],
 ['02:00', 6.746478873239437],
 ['13:00', 5.5777777777777775]]

** Top 5 Hours for Ask Posts Comments at Greenwich Mean Time **

In [17]:
for row in gmt_time_sorted[:5]:
    gmt = row[0] 
    gmt = dt.datetime.strptime(gmt,"%H:%M")
    gmt = dt.datetime.strftime(gmt, "%H:%M")
    avg_comm = row[1]
    string_3 = "{} {:.2f} average comments per post"
    output_1 = string_3.format(gmt, avg_comm)
    print(output_1)

19:00 38.59 average comments per post
06:00 23.81 average comments per post
00:00 21.52 average comments per post
20:00 16.80 average comments per post
01:00 16.01 average comments per post


### Findings 

1. The hour that receives the most comments per post on average is **19:00 GMT**, with an average of **38.59** comments per post. There's about a **60% increase** in the number of comments between the hours with the highest and second highest average number of comments. 

Next, we'll determine if show or ask posts receive more points on average. We'll use the following steps to perform this analysis:

1. Suming the number of points for the show posts and ask posts seperately. 
2. Average the points for both categories

In [18]:
# Summing and calculating average for show posts

show_post_points = 0

for row in show_posts:
    show_points = int(row[3])
    show_post_points += show_points
    
avg_show_post_points = show_post_points / len(show_posts)

print(show_post_points)
print(avg_show_post_points)

32019
27.555077452667813


In [19]:
# Summing and calculating average for ask posts

ask_post_points = 0

for row in ask_posts:
    ask_points = int(row[3])
    ask_post_points += ask_points
    
avg_ask_post_points = ask_post_points / len(ask_posts)

print(ask_post_points)
print(avg_ask_post_points)

26268
15.061926605504587


From the data, the it has been determined that:

1.  Show posts have **32,019** points whereas ask posts received **26,268** points. Thus show posts received more points than ask posts

2. Show posts received more points on average as compared to ask posts. Show posts averaged **27.55** points as compared to ask posts lower average of **15.06**

The next step is to compare the results of the average number of comments and points other posts receive to the show posts and ask posts.

In [20]:
# Summing and calculating average points for other posts

other_post_points = 0

for row in other_posts:
    other_points = int(row[3])
    other_post_points += other_points
    
avg_other_post_points = other_post_points / len(other_posts)

print(other_post_points)
print(avg_other_post_points)

952664
55.4067698034198


The other posts on the website received more (**952,664**) points than the ask and show posts combined. It also had a higher average number of points (**55.41**) than both categories.

In [21]:
# Summing and calculating average comments for other posts

other_post_comments = 0

for row in other_posts:
    other_comments = int(row[4])
    other_post_comments += other_comments
    
avg_other_post_comments = other_post_comments / len(other_posts)

print(other_post_comments)
print(avg_other_post_comments)

462055
26.8730371059672


It has been determined fro the data that: 

1. The other posts category received **462,055** comments for **17,194** posts. 

2. The other category received an average of **26.87** comments per post which is more than the **14.03** and **10.32** average number of comments recived by the ask post and show post respectively. 

## Conclusion

1. It was determined that the hour that receives the most comments per post on average is **19:00 GMT**, with an average of **38.59** comments per post.

2. It was also determined that the ask post received more comments than the show posts and also had a higher average number of comments per post. **14.03** for ask posts as compared to **10.32** for show posts

3. The show posts have more points than ask posts. The show posts got **32,019** points whereas ask posts received **26,268 points**. Thus show posts received more points than ask posts

4. Show posts received more points on average as compared to ask posts. Show posts averaged **27.55** points as compared to ask posts lower average of **15.06**

5. The other posts category received **462,055** comments for **17,194** posts. 

6. The other category received an average of **26.87** comments per post which is more than the **14.03** and **10.32** average number of comments recived by the ask post and show post respectively. 
